core/jobs/batch_jobs/contributor_admin_stats_jobs.py

```diff
@@ -0,0 +1,1236 @@
+                )
+            )
+            question_submit_stats_models.update_timestamps()
+            return question_submit_stats_models
+
+    @staticmethod
+    def transform_question_review_stats(
+        reviewer_user_id: str,
+        question_reviewer_stats:
+            Iterable[suggestion_models.QuestionReviewStatsModel]) -> (
+        suggestion_models.QuestionReviewerTotalContributionStatsModel):
+        """Transforms QuestionReviewStatsModel to
+        QuestionReviewerTotalContributionStatsModel.
+
+        Args:
+            reviewer_user_id: str. User ID acting as a key to new model.
+            question_reviewer_stats:
+                Iterable[suggestion_models.QuestionReviewStatsModel].
+                QuestionReviewStatsModel grouped by
+                reviewer_user_id.
+
+        Returns:
+            suggestion_models.QuestionReviewerTotalContributionStatsModel.
+            New QuestionReviewerTotalContributionStatsModel model.
+        """
+
+        question_reviewer_stats = list(question_reviewer_stats)
+        entity_id = reviewer_user_id
+
+        topic_ids = (
+            [v.topic_id for v in question_reviewer_stats])
+        reviewed_questions_count = sum(
+            [v.reviewed_questions_count
+                for v in question_reviewer_stats])
+        accepted_questions_count = sum(
+            [v.accepted_questions_count
+                for v in question_reviewer_stats])
+        accepted_questions_with_reviewer_edits_count = sum(
+            [v.accepted_questions_with_reviewer_edits_count
+                for v in question_reviewer_stats])
+        rejected_questions_count = (
+            reviewed_questions_count - accepted_questions_count
+        )
+        first_contribution_date = min(
+            [v.first_contribution_date for v in question_reviewer_stats])
+        last_contribution_date = max(
+            [v.last_contribution_date for v in question_reviewer_stats])
+
+        with datastore_services.get_ndb_context():
+            question_review_stats_models = (
+                suggestion_models.QuestionReviewerTotalContributionStatsModel(
+                id=entity_id,
+                contributor_id=reviewer_user_id,
+                topic_ids_with_question_reviews=topic_ids,
+                reviewed_questions_count=reviewed_questions_count,
+                accepted_questions_count=accepted_questions_count,
+                accepted_questions_with_reviewer_edits_count=(
+                    accepted_questions_with_reviewer_edits_count),
+                rejected_questions_count=rejected_questions_count,
+                first_contribution_date=first_contribution_date,
+                last_contribution_date=last_contribution_date
+                )
+            )
+            question_review_stats_models.update_timestamps()
+            return question_review_stats_models
+
+
+class AuditGenerateContributorAdminStatsJob(base_jobs.JobBase):
```

Comment:
Can we avoid adding all the duplicate codes here?

Possible way:

```
class ActualJob:
    DATASTORE_UPDATES_ALLOWED = True
    def run():
        .....
        .....
        if DATASTORE_UPDATES_ALLOWED:
            <Steps calling ndb_io.PutModels()>
            
class AuditActualJob:
    DATASTORE_UPDATES_ALLOWED = False
```

@vojtechjelinek Do we have any official way to write audit jobs without duplicate code?

Reply:
Hi @DubeySandeep thanks for your suggestion but I am little confused on how it should be implemented cause the run method of ActualJob takes `self` as param. Can you please clarify a little bit more?

Reply:
@kshitij01042002 Sorry for the confusion, the above code is for reference and doesn't have the correct code. You can check [this existing job for help](https://github.com/oppia/oppia/blob/5957b9ddc0b58aa61477a3685195b39ac015dfe4/core/jobs/batch_jobs/translation_migration_jobs.py#L195).

Reply:
Thanks @DubeySandeep for the reference made the suggested changes PTAL.

Reply:
@DubeySandeep I already responded somewhere else to this, there isn't really a standardized way to do this, it needs to be thought through.